src/Cassandra.IntegrationTests/MetadataTests/TokenMapSchemaChangeTests.cs

```diff
@@ -0,0 +1,93 @@
+        {
+        }
+
+        [Test]
+        public void TokenMap_Should_UpdateExistingTokenMap_When_KeyspaceIsCreated()
+        {
+            var keyspaceName = TestUtils.GetUniqueKeyspaceName().ToLower();
+            Assert.Throws<KeyNotFoundException>(() => Cluster.Metadata.TokenToReplicasMap.GetByKeyspace(keyspaceName));
+            var oldTokenMap = Cluster.Metadata.TokenToReplicasMap;
+            var createKeyspaceCql = $"CREATE KEYSPACE {keyspaceName} WITH replication = {{'class': 'SimpleStrategy', 'replication_factor' : 3}}";
+            Assert.AreEqual(3, Cluster.Metadata.Hosts.Count);
+
+            Session.Execute(createKeyspaceCql);
+            TestUtils.WaitForSchemaAgreement(Cluster);
+            Session.ChangeKeyspace(keyspaceName);
+
+            TestHelper.RetryAssert(() =>
+            {
+                IReadOnlyDictionary<IToken, ISet<Host>> replicas = null;
+                Assert.DoesNotThrow(() =>
+                {
+                    replicas = Cluster.Metadata.TokenToReplicasMap.GetByKeyspace(keyspaceName);
+                });
+                Assert.AreEqual(Cluster.Metadata.Hosts.Sum(h => h.Tokens.Count()), replicas.Count);
+            });
+            Assert.IsTrue(object.ReferenceEquals(Cluster.Metadata.TokenToReplicasMap, oldTokenMap));
+        }
+
+        [Test]
+        public void TokenMap_Should_UpdateExistingTokenMap_When_KeyspaceIsRemoved()
+        {
+            var oldTokenMap = Cluster.Metadata.TokenToReplicasMap;
+            var removeKeyspaceCql = $"DROP KEYSPACE {KeyspaceName}";
+            Session.Execute(removeKeyspaceCql);
+            TestUtils.WaitForSchemaAgreement(Cluster);
+
+            TestHelper.RetryAssert(() =>
+            {
+                Assert.Throws<KeyNotFoundException>(() => Cluster.Metadata.TokenToReplicasMap.GetByKeyspace(KeyspaceName));
+            });
+            Assert.IsTrue(object.ReferenceEquals(Cluster.Metadata.TokenToReplicasMap, oldTokenMap));
+        }
+
+        [Test]
+        public void TokenMap_Should_UpdateExistingTokenMap_When_KeyspaceIsChanged()
+        {
+            Assert.AreEqual(3, Cluster.Metadata.Hosts.Count);
+            var oldTokenMap = Cluster.Metadata.TokenToReplicasMap;
+            var alterKeyspaceCql = $"ALTER KEYSPACE {KeyspaceName} WITH replication = {{'class': 'SimpleStrategy', 'replication_factor' : 3}}";
+            Session.Execute(alterKeyspaceCql);
+            TestUtils.WaitForSchemaAgreement(Cluster);
+            TestHelper.RetryAssert(() =>
+            {
+                var replicas = Cluster.Metadata.TokenToReplicasMap.GetByKeyspace(KeyspaceName);
+                Assert.AreEqual(Cluster.Metadata.Hosts.Sum(h => h.Tokens.Count()), replicas.Count);
+                Assert.AreEqual(3, Cluster.Metadata.GetReplicas(KeyspaceName, Encoding.UTF8.GetBytes("123")).Count);
+            });
+            Assert.IsTrue(object.ReferenceEquals(Cluster.Metadata.TokenToReplicasMap, oldTokenMap));
```

Comment:
I think the CI failures are due to the fact that this involves multiple threads updating a value (`Cluster.Metadata.TokenToReplicasMap`) which is volatile but the `oldTokenMap` write and reads have no memory barriers.

We should use `Volatile.Write()` and `Volatile.Read()` when accessing `oldTokenMap` to make sure that read/write instructions are not reordered. The same applies to the rest of the tests.

Reply:
There's only one write to `oldTokenMap` before the actual test begins, so I don't see the need to use `Volatile.Write()` and `Volatile.Read()`. There was only one test that was failing in CI which is `TokenMapUpdates_Should_FunctionCorrectly_When_MultipleThreadsCallingRebuildAndUpdateKeyspace` (unit test). It failed only once since the introduction of the `ticks` checks but it might have been because of the resolution like it was mentioned previously.

Reply:
That failure might be because those tests are sharing a `Cluster` instance. I've changed them to create their own instance.